File: nodeTOPFLYtechCodec/PTTDecoder.js
var CryptoTool = require("./CryptoTool")
var ByteUtils = require("./ByteUtils")
const TopflytechByteBuf = require("./TopflytechByteBuf")

var PTTDecoder = {
    HEADER_LENGTH:3,
    HEARTBEAT :  [0x28, 0x28, 0x03],
    TALK_START :  [0x28, 0x28, 0x04],
    TALK_END :  [0x28, 0x28, 0x05],
    VOICE_DATA :  [0x28, 0x28, 0x06],
    encryptType:0,
    aesKey:"",
    match:function (bytes){
        if(bytes.length != this.HEADER_LENGTH){
            return false
        }
        return ByteUtils.arrayEquals(this.HEARTBEAT, bytes)
            || ByteUtils.arrayEquals(this.TALK_START, bytes)
            || ByteUtils.arrayEquals(this.TALK_END, bytes)
            || ByteUtils.arrayEquals(this.VOICE_DATA, bytes) ;
    },

    decode(buf){
        TopflytechByteBuf.putBuf(buf);
        var messages = [];
        if (TopflytechByteBuf.getReadableBytes() < (this.HEADER_LENGTH + 2)){
            return messages;
        }
        var foundHead = false;
        var bytes= [3]
        while (TopflytechByteBuf.getReadableBytes() > 5){
            TopflytechByteBuf.markReaderIndex();
            bytes[0] = TopflytechByteBuf.getByte(0);
            bytes[1] = TopflytechByteBuf.getByte(1);
            bytes[2] = TopflytechByteBuf.getByte(2);
            if (this.match(bytes)){
                foundHead = true;
                TopflytechByteBuf.skipBytes(this.HEADER_LENGTH);
                var lengthBytes = TopflytechByteBuf.readBytes(2);
                var packageLength = ByteUtils.byteToShort(lengthBytes, 0);
                if (this.encryptType == CryptoTool.MessageEncryptType.MD5){
                    packageLength = packageLength + 8;
                }else if(this.encryptType == CryptoTool.MessageEncryptType.AES){
                    packageLength = CryptoTool.AES.getAesLength(packageLength);
                }
                TopflytechByteBuf.resetReaderIndex();
                if(packageLength <= 0){
                    TopflytechByteBuf.skipBytes(5);
                    break;
                }
                if (packageLength > TopflytechByteBuf.getReadableBytes()){
                    break;
                }
                var data = TopflytechByteBuf.readBytes(packageLength);
                data = CryptoTool.decryptData(data, this.encryptType, this.aesKey);
                if (data != null){
                    try {
                        var message = this.build(data);
                        if (message != null){
                            messages.push(message);
                        }
                    }catch (e){
                        console.log(e)
                    }
                }
            }else{
                TopflytechByteBuf.skipBytes(1);
            }
        }
        return messages
    },
    build:function(bytes){
        if (bytes != null && bytes.length > this.HEADER_LENGTH
            && ((bytes[0] == 0x28 && bytes[1] == 0x28))) {
            switch (bytes[2]) { 
                case 0x03:
                    var heartbeatMessage = this.parseHeartbeat(bytes);
                    return heartbeatMessage;
                case 0x04:
                    var talkStartMessage = parseTalkStartMessage(bytes);
                    return talkStartMessage;
                case 0x05:
                    var talkEndMessage = parseTalkEndMessage(bytes);
                    return talkEndMessage;
                case 0x06:
                    var voiceMessage = parseVoiceMessage(bytes);
                    return voiceMessage;
                default:
                    return null;
            }
        }
    },
    parseHeartbeat:function (bytes){
        var serialNo = ByteUtils.byteToShort(bytes,5);
        var imei = ByteUtils.IMEI.decode(bytes,7)
        var heartbeatMessage = {
            serialNo:serialNo,
            imei:imei,
            srcBytes:bytes,
            messageType:"heartbeat",
        }
        return heartbeatMessage
    },
    parseTalkStartMessage:function (bytes){
        var serialNo = ByteUtils.byteToShort(bytes,5);
        var imei = ByteUtils.IMEI.decode(bytes,7)
        var talkStartMessage = {
            serialNo:serialNo,
            imei:imei,
            srcBytes:bytes,
            messageType:"talkStart",
        }
        return talkStartMessage
    },
    parseTalkEndMessage:function (bytes){
        var serialNo = ByteUtils.byteToShort(bytes,5);
        var imei = ByteUtils.IMEI.decode(bytes,7)
        var talkEndMessage = {
            serialNo:serialNo,
            imei:imei,
            srcBytes:bytes,
            messageType:"talkEnd",
        }
        return talkEndMessage
    },
    parseVoiceMessage:function (bytes){
        var serialNo = ByteUtils.byteToShort(bytes,5);
        var imei = ByteUtils.IMEI.decode(bytes,7)
        var voiceMessage = {
            serialNo:serialNo,
            imei:imei,
            srcBytes:bytes,
            messageType:"talkVoice",
        }
        voiceMessage.encodeType = bytes[15]
        var voiceLen = ByteUtils.byteToShort(bytes,16);
        if(bytes.length >= 18 + voiceLen){
            var voiceData = ByteUtils.arrayOfRange(bytes,18,18 + voiceLen)
            voiceMessage.voiceData = voiceData 
        }
        return voiceMessage
    },
}


module.exports = PTTDecoder;

File: TopflytechCodec/C#/TopflytechCodec/TopflytechCodec/PTTDecoder.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using TopflytechCodec.Entities;

namespace TopflytechCodec
{
    public class PTTDecoder
    {
        private const int HEADER_LENGTH = 3;
        private static readonly byte[] HEARTBEAT = { 0x28, 0x28, 0x03 };
        private static readonly byte[] TALK_START = { 0x28, 0x28, 0x04 };
        private static readonly byte[] TALK_END = { 0x28, 0x28, 0x05 };
        private static readonly byte[] VOICE_DATA = { 0x28, 0x28, 0x06 };

        private readonly int encryptType;
        private readonly string aesKey;
        private TopflytechByteBuf decoderBuf = new TopflytechByteBuf();

        public PTTDecoder(int messageEncryptType, string aesKey)
        {
            this.encryptType = messageEncryptType;
            this.aesKey = aesKey;
        }

        public PTTDecoder(int messageEncryptType, string aesKey, int buffSize)
        {
            this.encryptType = messageEncryptType;
            this.aesKey = aesKey;
            this.decoderBuf = new TopflytechByteBuf(buffSize);
        }

        private static bool match(byte[] bytes)
        {
            Debug.Assert(bytes.Length >= HEADER_LENGTH, "command match: length is not 3!");
            return Utils.ArrayEquals(TALK_START, bytes)
                   || Utils.ArrayEquals(HEARTBEAT, bytes)
                   || Utils.ArrayEquals(TALK_END, bytes)
                   || Utils.ArrayEquals(VOICE_DATA, bytes);
        }

        public List<Message> decode(byte[] buf)
        {
            decoderBuf.PutBuf(buf);
            List<Message> messages = new List<Message>();
            if (decoderBuf.GetReadableBytes() < (HEADER_LENGTH + 2))
            {
                return messages;
            }

            byte[] bytes = new byte[3];
            while (decoderBuf.GetReadableBytes() > 5)
            {
                decoderBuf.MarkReaderIndex();
                bytes[0] = decoderBuf.GetByte(0);
                bytes[1] = decoderBuf.GetByte(1);
                bytes[2] = decoderBuf.GetByte(2);
                if (match(bytes))
                {
                    decoderBuf.SkipBytes(HEADER_LENGTH);
                    byte[] lengthBytes = decoderBuf.ReadBytes(2);
                    int packageLength = BytesUtils.Bytes2Short(lengthBytes, 0);
                    if (encryptType == MessageEncryptType.MD5)
                    {
                        packageLength += 8;
                    }
                    else if (encryptType == MessageEncryptType.AES)
                    {
                        packageLength = Crypto.GetAesLength(packageLength);
                    }
                    decoderBuf.ResetReaderIndex();
                    if (packageLength <= 0)
                    {
                        decoderBuf.SkipBytes(5);
                        break;
                    }
                    if (packageLength > decoderBuf.GetReadableBytes())
                    {
                        break;
                    }
                    byte[] data = decoderBuf.ReadBytes(packageLength);
                    data = Crypto.DecryptData(data, encryptType, aesKey);
                    if (data != null)
                    {
                        Message message = build(data);
                        if (message != null)
                        {
                            messages.Add(message);
                        }
                    }
                }
                else
                {
                    decoderBuf.SkipBytes(1);
                }
            }
            return messages;
        }

        public Message build(byte[] bytes)
        {
            if (bytes != null && bytes.Length > HEADER_LENGTH)
            {
                switch (bytes[2])
                {
                    case 0x03:
                        return parseHeartbeat(bytes);
                    case 0x04:
                        return parseTalkStartMessage(bytes);
                    case 0x05:
                        return parseTalkEndMessage(bytes);
                    case 0x06:
                        return parseVoiceMessage(bytes);
                    default:
                        return null;
                }
            }
            return null;
        }

        private VoiceMessage parseVoiceMessage(byte[] bytes)
        {
            int serialNo = BytesUtils.Bytes2Short(bytes, 5);
            string imei = BytesUtils.IMEI.Decode(bytes, 7);
            VoiceMessage voiceMessage = new VoiceMessage();
            voiceMessage.OrignBytes = bytes;
            voiceMessage.SerialNo = serialNo;
            voiceMessage.Imei = imei;
            voiceMessage.EncodeType = bytes[15];
            int voiceLen = BytesUtils.Bytes2Short(bytes, 16);
            if (bytes.Length >= 18 + voiceLen)
            {
                byte[] voiceData = Utils.ArrayCopyOfRange(bytes, 18, 18 + voiceLen);
                voiceMessage.VoiceData = voiceData;
            }
            return voiceMessage;
        }

        private TalkEndMessage parseTalkEndMessage(byte[] bytes)
        {
            int serialNo = BytesUtils.Bytes2Short(bytes, 5);
            string imei = BytesUtils.IMEI.Decode(bytes, 7);
            TalkEndMessage talkEndMessage = new TalkEndMessage();
            talkEndMessage.OrignBytes = bytes;
            talkEndMessage.SerialNo = serialNo;
            talkEndMessage.Imei = imei;
            return talkEndMessage;
        }

        private TalkStartMessage parseTalkStartMessage(byte[] bytes)
        {
            int serialNo = BytesUtils.Bytes2Short(bytes, 5);
            string imei = BytesUtils.IMEI.Decode(bytes, 7);
            TalkStartMessage talkStartMessage = new TalkStartMessage();
            talkStartMessage.OrignBytes = bytes;
            talkStartMessage.SerialNo = serialNo;
            talkStartMessage.Imei = imei;
            return talkStartMessage;
        }

        private HeartbeatMessage parseHeartbeat(byte[] bytes)
        {
            int serialNo = BytesUtils.Bytes2Short(bytes, 5);
            string imei = BytesUtils.IMEI.Decode(bytes, 7);
            HeartbeatMessage heartbeatMessage = new HeartbeatMessage();
            heartbeatMessage.OrignBytes = bytes;
            heartbeatMessage.SerialNo = serialNo;
            heartbeatMessage.Imei = imei;
            return heartbeatMessage;
        }
    }
}
